### PR TITLE
Add timeout handling.

### LINKED ISSUE #8: Implement timeout/deadline and cancelation

### PR DESCRIPTION
**Pull request by jakobr-google**
Fixes #8.

### CODE REVIEW

File: lib/src/shared.dart [hunk 1]
```diff
@@ -8,6 +8,57 @@ import 'package:async/async.dart';
...
+  }
+}
+
+/// Convert [timeout] from grpc-timeout header string format to [Duration].
+/// Returns [null] if [timeout] is not correctly formatted.
+Duration fromTimeoutString(String timeout) {
+  if (timeout == null) return null;
+  if (timeout.length < 2) return null;
+  try {
+    final value = int.parse(timeout.substring(0, timeout.length - 1));
```
Comment (johnniwinther) [inline lib/src/shared.dart]: Maybe use
```
final value = parse(..., onError: (_) => null);
if (value == null) return null;
```
to avoid exception handling?

File: test/timeout_test.dart [hunk 2]
```diff
@@ -2,24 +2,126 @@
...
-        ClientCall.toTimeoutString(new Duration(seconds: 2, microseconds: 3)),
-        '2000m');
-    expect(
-        ClientCall.toTimeoutString(new Duration(seconds: 2, milliseconds: 3)),
-        '2003m');
+  const dummyValue = 0;
+
+  group('Unit:', () {
+    test('Timeouts are converted correctly to header string', () {
+      expect(toTimeoutString(new Duration(microseconds: -1)), '1n');
```
Comment (johnniwinther) [inline test/timeout_test.dart]: Add test for `duration == null` and values in minutes and hours.

File: test/timeout_test.dart [hunk 3]
```diff
@@ -2,24 +2,126 @@
...
+          toTimeoutString(new Duration(hours: 2, microseconds: 17)), '7200S');
+      expect(toTimeoutString(new Duration(milliseconds: 1420665)), '1420S');
+      expect(
+          toTimeoutString(new Duration(seconds: 2, microseconds: 3)), '2000m');
+      expect(
+          toTimeoutString(new Duration(seconds: 2, milliseconds: 3)), '2003m');
+    });
+
+    test('Timeouts are converted correctly from header string', () {
+      expect(fromTimeoutString('1n'), new Duration(microseconds: 1000));
```
Comment (johnniwinther) [inline test/timeout_test.dart]: Add tests for `timeout == null`, timeouts in minutes and hours, and garbage data such as `1`, `1s` and `ab`, and interpreted values such as `-1S`.

File: lib/src/shared.dart (hunk 1 above)
Comment (jakobr-google) [inline lib/src/shared.dart]: Thanks! Done.

File: test/timeout_test.dart (hunk 2 above)
Comment (jakobr-google) [inline test/timeout_test.dart]: Done.

File: test/timeout_test.dart (hunk 3 above)
Comment (jakobr-google) [inline test/timeout_test.dart]: Done.
